File: benchmark/src/test/java/org/alveolo/ffm/benchmark/nativecall/AffmTest.java
package org.alveolo.ffm.benchmark.nativecall;

import java.lang.foreign.SegmentAllocator;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;

import org.alveolo.ffm.Address;
import org.alveolo.ffm.FirstVariadicArg;
import org.alveolo.ffm.ForeignInterface;
import org.alveolo.ffm.In;
import org.alveolo.ffm.Library;
import org.alveolo.ffm.Out;
import org.alveolo.ffm.Sequence;
import org.alveolo.ffm.SLong;
import org.alveolo.ffm.SizeT;
import org.alveolo.ffm.Symbol;
import org.alveolo.ffm.ULong;
import org.alveolo.ffm.Value;
import org.alveolo.ffm.WCharT;

@Library("affm_test")
@ForeignInterface
public interface AffmTest {
  ArrayVirtual get_array_virtual();

  VirtualPairs get_virtual_pairs();

  int add_ints(int left, int right);

  @Symbol("null_pointer") PairS absentPair();
  @Symbol("null_pointer") NullableUnion absentUnion();
  @Symbol("null_pointer") PairSFM absentWrapper();
  @Symbol("null_pointer") @Address PairR absentRecord();
  @Symbol("null_pointer") VirtualPairs absentVirtual();
  @Symbol("null_pointer") String absentString();
  @Symbol("null_pointer") @Address int absentInt();
  @Symbol("null_pointer") @Address @SizeT long absentSize();

  @Symbol("nullable_pair") int optionalPair(PairS value);
  @Symbol("nullable_pair") int optionalWrapper(PairSFM value);
  @Symbol("nullable_pair") int optionalRecord(@Address PairR value);
  @Symbol("nullable_string") int optionalString(String value);
  @Symbol("nullable_values")
  int optionalValues(@Address PairR value, String text);
  @Symbol("required_values")
  int requiredValues(PairR value, String text);
  @Symbol("nullable_values")
  int optionalInterfaceValues(PairS value, String text);
  @Symbol("nullable_box")
  int optionalAllocatingValues(@Address PairBoxRA value, String text);


  @SLong long echo_slong(@SLong long value);

  @ULong long echo_ulong(@ULong long value);

  @SizeT long echo_size_t(@SizeT long value);

  @SizeT long read_size_t(@Address @SizeT long value);

  @SizeT long sum_size_t(
      @Address @SizeT long first, @Address @SizeT long second);

  @Address @SizeT long size_t_address();

  @WCharT int echo_wchar(@WCharT int value);

  @Symbol("read_c_long")
  @SLong long read_c_long(@Address @SLong long value);

  @Address @SLong long c_long_address();

  @FirstVariadicArg(1)
  int variadic_sum(int count);

  @FirstVariadicArg(1)
  int variadic_sum(int count, int first, int second, int third);

  long utf8_bytes(String value);

  int set_errno_and_return(ErrnoSpec capture, int value, int error);

  default int checked_errno_return(
      ErrnoSpec capture, int value, int error) {
    int result = set_errno_and_return(capture, value, error);
    capture.throwIf(() -> result == -1);
    return result;
  }

  @Symbol("make_pair")
  PairR make_pair_record(int left, int right);

  PairR make_pair_and_set_errno(
      ErrnoSpec capture, int left, int right, int error);

  int pair_sum(PairR value);

  @Symbol("pair_ptr_sum")
  int pair_ptr_sum_record(@Address PairR ref);

  @Symbol("make_pair")
  @Value
  PairS make_pair(
      SegmentAllocator allocator, int left, int right);

  @Symbol("pair_sum")
  int pair_sum_interface_value(@Value PairS value);

  @Symbol("pair_ptr_sum")
  int pair_ptr_sum_interface(PairS ref);

  @Symbol("pair_box_value_sum")
  int pair_box_record_value_sum(PairBoxRV value);

  @Symbol("pair_box_ptr_sum")
  int pair_box_record_address_sum(PairBoxRA value);

  @Symbol("pair_box_ptr_sum")
  int pair_box_interface_address_sum(@Value PairBoxIA value);

  @Symbol("pair_box_value_sum")
  int pair_box_interface_value_sum(@Value PairBoxIV value);

  void scale_ints(
      int[] values, int count, int factor);

  default void scale_ints(int[] values, int factor) {
    scale_ints(values, values.length, factor);
  }

  void offset_pairs(
      PairR[] values, int count, int delta);

  default void offset_pairs(PairR[] values, int delta) {
    offset_pairs(values, values.length, delta);
  }

  void fill_pairs(
      @Out PairR[] values, int count, int start);

  default void fill_pairs(PairR[] values, int start) {
    fill_pairs(values, values.length, start);
  }

  void mutate_native_arrays(NativeArrays value);

  @Symbol("sum_three_and_clobber")
  int sum_three_and_clobber(@In @Sequence(3) int[] values);

  @Symbol("fill_two_ints")
  void fill_two_ints(@Out @Sequence(2) int[] values, int start);

  void increment_bytes(
      ByteBuffer values, int count);

  default void increment_bytes(ByteBuffer values) {
    increment_bytes(values, values.remaining());
  }

  @Symbol("fill_two_ints")
  void fill_two_int_buffer(@Out @Sequence(2) IntBuffer values, int start);

  @Symbol("scale_ints")
  void scale_int_buffer(
      IntBuffer values, int count, int factor);

  default void scale_int_buffer(IntBuffer values, int factor) {
    scale_int_buffer(values, values.remaining(), factor);
  }

  @Symbol("sum_three_and_clobber")
  int sum_three_int_buffer(@In @Sequence(3) IntBuffer values);

  int sum_int3_value(@Value @Sequence(3) int[] values);

  @Symbol("sum_int3_value")
  int sum_int3_buffer_value(@Value @Sequence(3) IntBuffer values);
}

File: benchmark/src/test/resources/native/affm_test.c
#include <stdint.h>
#include <stddef.h>
#include <errno.h>
#include <stdarg.h>
#include <wchar.h>

#ifdef _WIN32
#define EXPORT __declspec(dllexport)
#else
#define EXPORT
#endif

typedef struct pair {
  int left;
  int right;
} pair;

typedef struct pair_box_value {
  pair pair;
} pair_box_value;

typedef struct pair_box_ptr {
  pair* pair;
} pair_box_ptr;

typedef struct native_arrays {
  int32_t matrix[2][3];
  pair points[2];
  pair* pointers[2];
} native_arrays;

typedef struct int3_value {
  int32_t values[3];
} int3_value;

static void virtual_fill(void* self, int32_t* values, int32_t count) {
  for (int32_t i = 0; i < count; i++) {
    values[i] = 10 + i;
  }
}

// Count validation tests must remain safe even if Java validation regresses.
static int32_t virtual_count(void* self, int32_t* values, int32_t count) {
  return count;
}

typedef struct counted_vtable {
  void (*fill_array)(void*, int32_t*, int32_t);
  void (*fill_buffer)(void*, int32_t*, int32_t);
  int32_t (*count)(void*, int32_t*, int32_t);
} counted_vtable;

typedef struct counted_virtual {
  const counted_vtable* vtable;
} counted_virtual;

EXPORT const counted_virtual* get_counted_virtual(void) {
  static const counted_vtable vtable = {
    virtual_fill, virtual_fill, virtual_count
  };
  static const counted_virtual object = {&vtable};
  return &object;
}

EXPORT int add_ints(int left, int right) {
  return left + right;
}

EXPORT long echo_slong(long value) {
  return value;
}

EXPORT unsigned long echo_ulong(unsigned long value) {
  return value;
}

EXPORT size_t echo_size_t(size_t value) {
  return value;
}

EXPORT size_t read_size_t(const size_t* value) {
  return *value;
}

EXPORT size_t sum_size_t(const size_t* first, const size_t* second) {
  return *first + *second;
}

EXPORT const size_t* size_t_address(void) {
  static const size_t value = (size_t)-1;
  return &value;
}

EXPORT wchar_t echo_wchar(wchar_t value) {
  return value;
}

EXPORT long read_c_long(const long* value) {
  return *value;
}

EXPORT long* c_long_address(void) {
  static long value = 123;
  return &value;
}

EXPORT int variadic_sum(int count, ...) {
  va_list arguments;
  va_start(arguments, count);

  int result = 0;
  for (int i = 0; i < count; i++) {
    result += va_arg(arguments, int);
  }

  va_end(arguments);
  return result;
}

EXPORT int set_errno_and_return(int value, int error) {
  errno = error;
  return value;
}

EXPORT int64_t utf8_bytes(const char* value) {
  int64_t count = 0;
  while (value[count] != '\0') {
    count++;
  }
  return count;
}

EXPORT pair make_pair(int left, int right) {
  pair result = {left, right};
  return result;
}

EXPORT pair make_pair_and_set_errno(int left, int right, int error) {
  errno = error;
  pair result = {left, right};
  return result;
}

static pair virtual_make_pair(void* self, int left, int right) {
  return make_pair(left, right);
}

static pair virtual_make_pair_with_error(
    void* self, int left, int right, int error) {
  return make_pair_and_set_errno(left, right, error);
}

typedef struct virtual_pairs_vtable {
  pair (*make)(void*, int, int);
  pair (*make_with_error)(void*, int, int, int);
} virtual_pairs_vtable;

typedef struct virtual_pairs {
  const virtual_pairs_vtable* vtable;
} virtual_pairs;

EXPORT const virtual_pairs* get_virtual_pairs(void) {
  static const virtual_pairs_vtable vtable = {
    virtual_make_pair, virtual_make_pair_with_error
  };
  static const virtual_pairs object = {&vtable};
  return &object;
}

EXPORT int pair_sum(pair value) {
  return value.left + value.right;
}

EXPORT int pair_ptr_sum(const pair* value) {
  return value->left + value->right;
}

EXPORT int pair_box_value_sum(pair_box_value value) {
  return pair_sum(value.pair);
}

EXPORT int pair_box_ptr_sum(pair_box_ptr value) {
  return pair_ptr_sum(value.pair);
}

EXPORT void scale_ints(int32_t* values, int32_t count, int32_t factor) {
  for (int32_t i = 0; i < count; i++) {
    values[i] *= factor;
  }
}

EXPORT void offset_pairs(pair* values, int32_t count, int32_t delta) {
  for (int32_t i = 0; i < count; i++) {
    values[i].left += delta;
    values[i].right += delta;
  }
}

EXPORT void fill_pairs(pair* values, int32_t count, int32_t start) {
  for (int32_t i = 0; i < count; i++) {
    values[i].left = start + i * 2;
    values[i].right = start + i * 2 + 1;
  }
}

EXPORT void mutate_native_arrays(native_arrays* values) {
  values->matrix[1][2] += 100;
  values->points[1].left += 10;
  values->points[1].right += 20;
  if (values->pointers[0] != 0) {
    values->pointers[0]->left += 30;
    values->pointers[0]->right += 40;
  }
  if (values->pointers[1] == 0) {
    values->matrix[0][1] = 77;
  }
}

EXPORT int32_t sum_three_and_clobber(int32_t* values) {
  int32_t result = values[0] + values[1] + values[2];
  values[0] = 777;
  return result;
}

EXPORT void fill_two_ints(int32_t* values, int32_t start) {
  values[0] = start;
  values[1] = start + 1;
}

EXPORT void increment_bytes(uint8_t* values, int32_t count) {
  for (int32_t i = 0; i < count; i++) {
    values[i] += 1;
  }
}

EXPORT int32_t sum_int3_value(int3_value value) {
  int32_t result = value.values[0] + value.values[1] + value.values[2];
  value.values[0] = 777;
  return result;
}
